### PR TITLE
feat(rnd): Introduce Sub-Graph on Agent Server

### PR DESCRIPTION
**Pull request by majdyz**
### Background

This change brings the capability to decompose a graph into sub-graphs. The objective of this feature is to allow a user to build a visually modular, and easier-to-understand graph. Also, allowing you to import a graph into your existing graph, without decluttering your existing graph.

This feature will require more implementation on the UI side, to allow the grouping of subgraph to be represented as a node in the builder.

### Changes 🏗️

Introduced a subgraph functionality with the following property:

* Sub-graph is simply a set of nodes that are grouped together, making it representable as a node.
* Sub-graph input & output pins/schema are the `InputBlock` / `OutputBlock` nodes present in the subgraph.
* The previous point implies that connecting two nodes from different sub-graphs, other than input/output nodes, is not allowed.
* Graph can be nested, but defined flatly, e.g.: graph is now only represented by three components: nodes, links, and subgraphs (a set of list of nodes). A nested subgraph is simply connecting a node inside a subgraph into another `InputBlock` node of another subgraph.

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

### CODE REVIEW

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 1]
```diff
@@ -126,65 +165,100 @@ def sanitize(name):
...
+
+                sanitized_name = sanitize(name)
+                if i == 0:
+                    fields = block.output_schema.get_fields()
+                else:
+                    fields = block.input_schema.get_fields()
+                if sanitized_name not in fields:
+                    raise ValueError(f"{suffix}, `{name}` invalid, fields: {fields}")
+
+            if (
```
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/data/graph.py:213]: Basically this is the main change: 
You can now group nodes into a subgraph, and you are not allowed to connect nodes from the different subgraph.

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 2]
```diff
@@ -277,7 +351,17 @@ async def get_graph_all_versions(graph_id: str) -> list[Graph]:
 
 
 async def create_graph(graph: Graph) -> Graph:
-    await AgentGraph.prisma().create(
+    async with transaction() as tx:
```
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/data/graph.py:362]: Graph creation involved many inserts, this definitely requires a transaction.

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 3]
```diff
@@ -290,11 +374,30 @@ async def create_graph(graph: Graph) -> Graph:
 
     await asyncio.gather(
         *[
-            AgentNode.prisma().create(
+            AgentGraph.prisma(tx).create(
```
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/data/graph.py:385]: First persist the actual graph, then persist all of it's subgraph.
subgraph is simply a logical grouping, so there will be no nested subgraph, even if the graph is actually nested.

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 4]
```diff
@@ -126,65 +173,100 @@ def sanitize(name):
...
+                if not block:
+                    raise ValueError(f"{suffix}, {node.block_id} is invalid block.")
+
+                sanitized_name = sanitize(name)
+                if i == 0:
+                    fields = block.output_schema.get_fields()
+                else:
+                    fields = block.input_schema.get_fields()
+                if sanitized_name not in fields:
+                    raise ValueError(f"{suffix}, `{name}` invalid, fields: {fields}")
```
Comment (ntindle) [inline rnd/autogpt_server/autogpt_server/data/graph.py:211]: nit: can we functionize some of this for clarity?

**Review by Swiftyos — APPROVED** [review]
A nice elegant solution to this problem.

File: rnd/autogpt_server/autogpt_server/data/graph.py (hunk 4 above)
Comment (majdyz) [reply rnd/autogpt_server/autogpt_server/data/graph.py:211]: Sure, I'll do a cleanup follow-up PR, will do that there

**Pull request merged**